### PR TITLE
Beam search

### PR DESCRIPTION
**Pull request by uralik**
There is Beam class now which keeps all the information about beam and can also dump nice graphs for you if you want to analyze different paths. It also supports extraction of nbest list of hypothesis and is doing length penalty as it was suggested by Google NMT paper.

This beam is intended to use with mini-batches, so its quite fast. The class structure inspired by OpenNMT project, though it is not that powerful as there.

It is model/agent agnostic , already works good for my internal agent. This is dump graph example:

In the following PR I will put changes to make it work in seq2seq agent.

### CODE REVIEW

**Comment by stephenroller** [comment]
Super cool dialog tree

File: parlai/core/beam.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        self.min_n_best = min_n_best
+
+    def get_output_from_current_step(self):
+        return self.outputs[-1]
+
+    def get_backtrack_from_current_step(self):
+        return self.bookkeep[-1]
+
+    def advance(self, softmax_probs):
+        voc_size = softmax_probs.size(-1)
```
Comment (alexholdenmiller) [inline parlai/core/beam.py]: can you add a comment on the dimensions of softmax probs?

File: parlai/core/beam.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            if self.eos_top_ts is None:
+                self.eos_top_ts = len(self.outputs) - 1
+
+    def done(self):
+        return self.eos_top and self.n_best_counter >= self.min_n_best
+
+    def get_top_hyp(self):
+        """
+        Helper function to get single best hypothesis
+        :return: hypothesis sequence represented as t{}-w{}-sc{:.{prec}f}
```
Comment (alexholdenmiller) [inline parlai/core/beam.py]: what does `t{}-w{}-sc{:.{prec}f}` mean?

File: parlai/core/beam.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        tok_id = best_idxs % voc_size
+
+        self.outputs.append(tok_id)
+        self.bookkeep.append(hyp_id)
+
+        #  check new hypos for eos label, if we have some, add to finished
+        for i in range(self.beam_size):
+            if self.outputs[-1][i] == self.eos:
+                #  this is finished hypo, adding to finished
+                self.finished.append((self.scores[i], len(self.outputs)-1, i))
```
Comment (alexholdenmiller) [inline parlai/core/beam.py]: could you actually switch these to a `namedtuple` (`Hypothesis`?)

File: parlai/core/beam.py (hunk 1 above)
Comment (uralik) [reply parlai/core/beam.py]: done

File: parlai/core/beam.py (hunk 2 above)
Comment (uralik) [reply parlai/core/beam.py]: nvm, it does not exist anymore

**Pull request merged**